---
permalink: /end/
sidemenu: true
description: "Documentation for Monocle 3, getting started section."
modified: 2019-06-07
tags: [monocle, manual, vignette]
---

<div class= "container">
<div class= "row">

<div class= "col-sm-4">
  <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
</div>

<!-- main content area -->
<div class= "col-sm-8">


  <div class="btn-group">
    <button onClick="window.location.href = '{{site.baseurl}}/getting_started/';">Getting Started</button>
    <button onClick="window.location.href = '{{site.baseurl}}/cluster/';">Clustering</button>
    <button onClick="window.location.href = '{{site.baseurl}}/trajectory/';">Constructing Trajectories</button>
    <button onClick="window.location.href = '{{site.baseurl}}/differential_expression_analysis/';">Differential Expression Analysis</button>
    <button class="chosen" onClick="window.location.href = '{{site.baseurl}}/end/';">Citations</button>
  </div>
  

  <!--
  <h2>Appendices</h2>

  <h2>Theory Behind Monocle </h2>

  <h3>Reversed graph embedding </h3>

  <p> Single-cell expression datasets are some of the largest and most complex encountered
  in genomics. Even the smallest single-cell RNA-Seq experiments sample hundreds of cells, measuring the
  expression level of the more than 20,000 genes in each cell. Visualizing these
  datasets, identifying cells of different types, and comparing them to one another
  all pose major bioinformatics challenges. </p>

  <p> <em>Manifold learning</em> is a common strategy for dealing with complex, high-dimensional
  data. The premise of this approach is simple: the data may reside in a very high-dimensional space,
  but the intrinsic structure of the dataset is much simpler. Moreover, the data
  are not random - they are generated by a process that can be understood by
  inspecting the global structure of the dataset. For example, a single single-cell RNA-Seq
  experiment may reside in 20,000 dimensions, but the cells might all lie on or "near"
  a curve <em>embedded</em> within a much lower dimensional space. For example, we
  might expect that cells in different phases of the cell cycle be distributed along
  a closed loop. Indeed a recent large-scale single-cell RNA-Seq study found exactly
  that <a href= "https://www.ncbi.nlm.nih.gov/pubmed/26000488">[11]</a>. </p>

  <p>Manifold learning often involves <em>dimensionality reduction</em> techniques as a
  first step. Conventional dimensionality reduction approaches (for
  example, PCA, ICA, Isomap, LLE, etc.) are limited in their ability to explictly
  recover the intrinisic structure from the data. </p>

  <p>Monocle 2 uses a technique called <em>reversed graph embedding</em> <a href= "http://www.biorxiv.org/content/early/2017/02/21/110668">[10,</a> <a href= "http://ieeexplore.ieee.org/document/7769209/?reload=true">12]</a>  to learn the
  structure of the manifold that describes a single-cell experiment. It simultaneously: </p>

  <ol>
  <li> Reduces high dimensional expression data into a lower dimension space.
  <li> Learns an explicit, smooth manifold that generates the data.
  <li> Assigns each cell to its position on that manifold
  </ol>

  <p>Together, these tasks allow Monocle 2 to order cells in pseudotime in an entirely
  unsupervised, data-driven way. Importantly, Monocle 2 learns manifolds that are
  trees without needing any <em>a priori</em> information about the structure of the
  tree. Users do not need to provide Monocle 2 with constraints on the number of branches, etc.
  These are learned from the data. This allows Monocle 2 to to discriminate between
  linear and branched trajectories automatically. To our knowledge, Monocle 2 is the first
  trajectory reconstruction algorithm to learn smooth tree-like manifolds without
  needing to know its high-level structure ahead of time.

  <p>Reversed graph embedding simultaneously learns a <em>principal graph</em> that
  approximates the manifold, as well as a function that maps points on the graph
  (which is embedded in low dimensions) back to the original high dimensional space.
  Reversed graph embedding aims to learn both a set of <em>latent points</em>
  $$ \mathcal{Z} = \{\mathbf{z}_1, ..., \mathbf{z}_M\}$$corresponding to the input data
  that reside in the low-dimensional space along with a graph $\mathcal{G}$ that
  connects them. This graph approximates the manifold. In order to map points on
  the manifold back to the original high-dimensional input space, we also need to
  learn a function $f_{\mathcal{G}}$.

  <p>Learning a good reversed graph embedding can be described as an optimization problem that
  joint captures the positions of the latent points $\mathbf{z}$, the graph $\mathcal{G}$, and the function
  $f_\mathcal{G}$.

  <p>To learn the positions of the latent points $\mathbf{z}$, we must optimize:

  \begin{equation}
  \mathop{min}_{f_g \in \mathcal{F}} \mathop{min}_{\{\mathbf{z}_1, ...,
  \mathbf{z}_M\}} \sum_{i = 1}^N ||\mathbf{x}_i - f_g (\mathbf{z}_i)||^2
  \end{equation}

  <p>Given a set of latent point coordinates, the optimization of graph inference can
  be represented as:

  \begin{equation} \label{eq:mintree}
  \mathop{min}_{f_\mathcal{G} \in \mathcal{F}} \mathop{min}_{\{\mathbf{z}_1, ..., \mathbf{z}_M\}} \sum_{(V_i, V_j) \in
  \mathcal{E}} b_{i,j}||f_g(\mathbf{z}_i) - f_g(\mathbf{z}_j)||^2
  \end{equation}

  <p>where
  $\mathcal{X} = \{ \mathbf{x}_1, ..., \mathbf{x}_N\}$ are the original
  single-cell expression profiles. The $V_i$ are the the vertices of the
  undirected graph $\mathcal{G} = (\mathcal{V}, \mathcal{E})$. The weights for the
  edge in $\mathcal{E}$ are encoded as $b_{ij}$.

  <p>The first optimization problem aims to position the latent points such that their
  image under $f_\mathcal{G}$ (that is, their corresponding positions in the high-dimensional space)
  will be "close" to the input data. The second optimization aims to keep latent
  points that are close to one another in the low dimensional space close to one
  another in the high dimensional space as well. These two goals must be balanced
  against one another. Reversed graph embedding achieves this through the parameter
  $\lambda$

  \begin{equation}
  \mathop{min}_{\mathcal{G} \in \hat{\mathcal{G}}_b}\mathop{min}_{f_g \in \mathcal{F}} \mathop{min}_{\{\mathbf{z}_1, ...,
  \mathbf{z}_M\}} \sum_{i = 1}^N ||\mathbf{x}_i - f_g (\mathbf{z}_i)||^2 + \frac{\lambda}{2} \sum_{(V_i, V_j) \in \mathcal{E}}
  b_{i,j}||f_g(\mathbf{z}_i) - f_g(\mathbf{z}_j)||^2
  \end{equation}

  <p>Reversed graph embedding requires a feasible set $\hat{\mathcal{G}}_b$ of graphs and a
  mapping function $f_\mathcal{G}$. In practice, implementing reversed graph embedding
  requires that we place some constraints on $\hat{\mathcal{G}}_b$ and $f_\mathcal{G}$.
  As work on reversed graph embedding continues, we anticipate that more general
  schemes that consider a wider range of feasible graphs and mapping functions will
  become available. Monocle users should expect more general reversed graph embedding
  schemes in future versions.

  <p>Mao <em>et al</em> initially described two specific ways to implement the general framework
  of reversed graph embedding. Both are briefly summarized below. See the original
  paper on DDRTree for more details. Monocle 2 uses the second scheme, but can easily be
  run in a mode that corresponds to the first.
   -->


  <h2>Citation </h2>


  If you use Monocle to analyze your experiments, please cite:

  {% highlight R %}

  citation("monocle3")

  ## Cole Trapnell and Davide Cacchiarelli et al (2014): The dynamics and regulators of cell fate decisions are
  ## revealed by pseudo-temporal ordering of single cells. Nature Biotechnology
  ##
  ## Xiaojie Qiu et al (2017): Single-cell mRNA quantification and differential analysis with Census. Nature Methods
  ##
  ## Xiaojie Qiu et al (2017): Reverse graph embedding resolves complex single-cell developmental trajectories.
  ## Nature Methods
  ##
  ## Junyue Cao, Malte Spielmann et al (2019): The single-cell transcriptional landscape of mammalian organogenesis.
  ## Nature

  {% endhighlight %}


  <h2>Acknowledgements </h2>

  <p>Monocle was originally built by Cole Trapnell and Davide Cacchiarelli, with substantial design input from John Rinn and Tarjei Mikkelsen. We are grateful to Sharif Bordbar, Chris Zhu, Amy Wagers and the Broad RNAi platform for technical assistance, and Magali Soumillon for helpful discussions. Cole Trapnell was supported by a Damon Runyon Postdoctoral Fellowship. Davide Cacchiarelli was supported by a Human Frontier Science Program Fellowship. Cacchiarelli and Mikkelsen were also supported by the Harvard Stem Cell Institute. John Rinn was the Alvin and Esta Star Associate Professor. This work was supported by NIH grants 1DP2OD00667, P01GM099117, and P50HG006193-01. This work was also supported in part by the Single Cell Genomics initiative, a collaboration between the Broad Institute and Fluidigm Inc.</p>

  <p>Monocle versions 2 and 3 were developed by Cole Trapnell's lab. Significant portions were written by Xiaojie Qiu and Hannah Pliner. The work was supported by NIH grant 1DP2HD088158, the W. M. Keck Foundation, as well as an Alfred P. Sloan Foundation Research Fellowship. Hannah Pliner is supported by the <a href="https://brotmanbaty.org/">Brotman Baty Institute</a>.</p>

  <h2>References </h2>

  <ol>
    <li>Cole Trapnell*, Davide Cacchiarelli*, Jonna Grimsby, Prapti Pokharel, Shuqiang Li, Michael Morse, Niall J. Lennon, Kenneth J. Livak, Tarjei S. Mikkelsen, and John L. Rinn.<br/>
    <a href = "http://www.nature.com/nbt/journal/v32/n4/abs/nbt.2859.html">
    The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells</a>.<br/> Nature Biotechnology, 2014.<br/><br/> </li>

    <li>Junyue Cao*, Jonathan S. Packer*, Vijay Ramani, Darren A. Cusanovich, Chau Huynh, Riza Daza, Xiaojie Qiu, Choli Lee, Scott N. Furlan, Frank J. Steemers, Andrew Adey, Robert H. Waterston, Cole Trapnell**, Jay Shendure** .<br/>
    <a href = "http://http//science.sciencemag.org/content/357/6352/661">
    Comprehensive single-cell transcriptional profiling of a multicellular organism</a>.<br/>Science, 2017.<br/><br/>  </li>

    <li>Jacob H. Levine, Erin F. Simonds, Sean C. Bendall, Kara L. Davis, El-ad D. Amir, Michelle Tadmor, Oren Litvin, Harris Fienberg, Astraea Jager, Eli Zunder, Rachel Finck, Amanda L. Gedman, Ina Radtke, James R. Downing, Dana Pe’er** and Garry P. Nolan**.<br/>
      <a href = "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4508757/">
        Data-driven phenotypic dissection of AML reveals progenitor-like cells that correlate with prognosis</a>.<br/>Cell, 2015.<br/><br/>  </li>

    <li>F. Alexander Wolf,, Fiona K. Hamey, Mireya Plass, Jordi Solana, Joakim S. Dahlin, Berthold Göttgens, Nikolaus Rajewsky, Lukas Simon and Fabian J. Theis<br/>
      <a href = "https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">
        PAGA: graph abstraction reconciles clustering with trajectory inference through a topology preserving map of single cells</a>.<br/>Genome biology, 2018.<br/><br/>  </li>

    <li>Junyue Cao*, Malte Spielmann*, Xiaojie Qiu, Xingfan Huang, Daniel M. Ibrahim, Andrew J. Hill, Fan Zhang, Stefan Mundlos, Lena Christiansen, Frank J. Steemers, Cole Trapnell**, and Jay Shendure** <br/>
      <a href = "http://dx.doi.org/10.1038/s41586-019-0969-x">
            The single-cell transcriptional landscape of mammalian organogenesis</a>.<br/>Nature, 2019.<br/><br/>  </li>

    <li>Jonathan S. Packer*, Qin Zhu*, Chau Huynh, Priya Sivaramakrishnan, Elicia Preston, Hannah Dueck, Derek Stefanik, Kai Tan, Cole Trapnell, Junhyong Kim**, Robert H. Waterston**, John I. Murray** <br/>
      <a href = "http://dx.doi.org/10.1101/565549">
        A lineage-resolved molecular atlas of C. elegans embryogenesis at single cell resolution</a>.<br/>bioRxiv, 2019.<br/><br/>  </li>

    <li>Nadav Sharon*, Raghav Chawla*, Jonas Mueller, Jordan Vanderhooft, Luke James Whitehorn, Benjamin Rosenthal, Mads Gurtler, Ralph R. Estanboulieh, Dmitry Shvartsman, David K. Gifford, Cole Trapnell** and Doug Melton**<br/>
          <a href = "http://dx.doi.org/10.1016/j.cell.2018.12.003">
            A peninsular structure coordinates asynchronous differentiation with morphogenesis to generate pancreatic islets</a>.<br/>Cell, 2019.<br/><br/>  </li>

    <li>Xiaojie Qiu, Qi Mao, Ying Tang, Li Wang, Raghav Chawla, Hannah Pliner, and Cole Trapnell<br/>
      <a href = "http://dx.doi.org/doi:10.1038/nmeth.4402">
        Reversed graph embedding resolves complex single-cell developmental trajectories</a>.<br/>Nature methods, 2017<br/><br/>  </li>

     <li>X Qiu, A Hill, J Packer, D Lin, YA Ma, and C Trapnell.<br/>
          <a href = "http://dx.doi.org/10.1038/nmeth.4150">
            Single-cell mrna quantification and differential analysis with census</a>.<br/>Nature methods, 2017<br/><br/>  </li>

  </ol>
  <!--
  <p> [1] Cole Trapnell, Davide Cacchiarelli, Jonna Grimsby, Prapti Pokharel, Shuqiang Li, Michael Morse, Niall J. Lennon, Kenneth J. Livak, Tarjei S. Mikkelsen, and John L. Rinn. <a href = "http://www.nature.com/nbt/journal/v32/n4/abs/nbt.2859.html"> The dynamics and regulators of cell fate decisions are revealed by pseudotemporal ordering of single cells</a>. Nature Biotechnology, 2014. </p>

  <p> [2] Cole Trapnell, Adam Roberts, Loyal Goff, Geo Pertea, Daehwan Kim, David R Kelley, Harold Pimentel, Steven L Salzberg, John L Rinn, and Lior Pachter. <a href = "https://www.ncbi.nlm.nih.gov/pubmed/22383036"> Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks</a>. Nature Protocols, 7(3):562–578, March 2012. </p>

  <p> [3] Laurens van der Maaten and Geoffrey Hinton. <a href = "http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf"> Visualizing data using t-SNE</a>. J. Mach. Learn. Res., 9(Nov):2579– 2605, 2008. </p>

  <p> [4] Alex Rodriguez. Alessandro Laio. <a href = "http://science.sciencemag.org/content/344/6191/1492"> Clustering by fast search and find of density peaks</a>. Science, 344:1492–1496, 2014. </p>

  <p> [5] Hadley Wickham. <a href = "http://www.springer.com/us/book/9780387981413"> ggplot2: Elegant Graphics for Data Analysis</a>. Springer-Verlag New York, 2009. </p>

  <p> [6] Simon Anders and Wolfgang Huber. <a href = "https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106"> Differential expression analysis for sequence count data</a>. Genome Biol., 11(10):R106, 2010. </p>

  <p> [7] Qi Mao, Li Wang, Steve Goodison, and Yijun Sun. <a href= "http://dl.acm.org/citation.cfm?id= 2783309"> Dimensionality reduction via graph structure learning</a>. In Proceedings of the 21th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, pages 765–774. ACM, 2015. </p>

  <p> [8] X Qiu, A Hill, J Packer, D Lin, YA Ma, and C Trapnell. <a href= "https://www.ncbi.nlm.nih.gov/pubmed/28114287"> Single-cell mrna quantification and differential analysis with census</a>. Nature methods, 2017. </p>

  <p> [9] Rahul Satija, Jeffrey A Farrell, David Gennert, Alexander F Schier, and Aviv Regev. <a href= "https://www.ncbi.nlm.nih.gov/pubmed/25867923"> Spatial reconstruction of single-cell gene expression data</a>. Nat. Biotechnol., 33(5):495–502, May 2015. </p>

  <p> [10] Xiaojie Qiu, Qi Mao, Ying Tang, Li Wang, Raghav Chawla, Hannah Pliner, and Cole Trapnell. <a href= "http://www.biorxiv.org/content/early/2017/02/21/110668"> Reversed graph embedding resolves complex single-cell developmental trajectories</a>. 21 February 2017. </p>

  <p> [11] Evan Z Macosko, Anindita Basu, Rahul Satija, James Nemesh, Karthik Shekhar, Melissa Goldman, Itay Tirosh, Allison R Bialas, Nolan Kamitaki, Emily M Martersteck, John J Trombetta, David A Weitz, Joshua R Sanes, Alex K Shalek, Aviv Regev, and Steven A McCarroll. <a href= "https://www.ncbi.nlm.nih.gov/pubmed/26000488">Highly parallel genome-wide expression profiling of individual cells using nanoliter droplets</a>. Cell, 161(5):1202–1214, 2015. </p>

  <p> [12] Qi Mao, Li Wang, Ivor Tsang, and Yijun Sun. <a href= "http://ieeexplore.ieee.org/document/7769209/?reload=true">Principal graph and structure learning based on reversed graph embedding </a>. IEEE Trans. Pattern Anal. Mach. Intell., 5 December 2016. </p>

  <p> [13] Cabili, Moran N. et al. <a href = "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3185964/">Integrative Annotation of Human Large Intergenic Noncoding RNAs Reveals Global Properties and Specific Subclasses</a>. Genes & Development 25.18 (2011): 1915–1927. PMC. Web. 21 June 2017. </p>
   -->

   <div class="btn-toolbar">
     <button class="button other-button" onClick="window.location.href = '{{site.baseurl}}/differential_expression_analysis/';">&laquo; Previous</button>
   </div>

</div>
</div>
</div>
