---
permalink: /getting_started/
sidemenu: true
description: "Documentation for Monocle 3, getting started section."
modified: 2019-06-07
tags: [monocle, manual, vignette]
---

<div class= "container">
<div class= "row">
<div class= "col-sm-4">
  <nav id= "toc" data-spy= "affix" data-toggle= "toc"></nav>
</div>
<!-- main content area -->
<div class= "col-sm-8">

<div class="btn-group">
  <button class="chosen" onClick="window.location.href = '{{site.baseurl}}/getting_started/';">Getting Started</button>
  <button onClick="window.location.href = '{{site.baseurl}}/cluster/';">Clustering</button>
  <button onClick="window.location.href = '{{site.baseurl}}/trajectory/';">Constructing Trajectories</button>
  <button onClick="window.location.href = '{{site.baseurl}}/differential_expression_analysis/';">Differential Expression Analysis</button>
  <button onClick="window.location.href = '{{site.baseurl}}/end/';">Citations</button>
</div>

<h2>Introduction</h2>

<p>Single-cell transcriptome sequencing (sc-RNA-seq) experiments allow us to
  discover new cell types and help us understand how they arise in development.
  The <strong>Monocle 3</strong> package provides a toolkit for analyzing
  single-cell gene expression experiments.
</p>

<p>Monocle 3 can help you perform three main types of analysis:
<ul>
<li><strong>Clustering, classifying, and counting cells.</strong> Single-cell
  RNA-Seq experiments allow you to discover new (and possibly rare) subtypes of
  cells. Monocle 3 helps you identify them.</li>
<li><strong>Constructing single-cell trajectories.</strong> In development,
  disease, and throughout life, cells transition from one state to another.
  Monocle 3 helps you discover these transitions.</li>
<li><strong>Differential expression analysis.</strong> Characterizing new cell
  types and states begins with comparisons to other, better understood cells.
  Monocle 3 includes a sophisticated, but easy-to-use system for differential
  expression.</li>
</ul>
</p>

<div class= "panel panel-warning">
<div class= "panel-heading">
<h3 class= "panel-title" data-toc-skip>Under construction</h3>
</div>
<div class= "panel-body">
<p>Monocle 3 is currently in the
  <a href="https://en.wikipedia.org/wiki/Software_release_life_cycle#Beta">beta phase</a>
  of its development. This means there are likely bugs and performance issues
  that will need to be addressed. We are working hard towards a stable release,
  but please be patient while Monocle 3 is under construction.</p>

<p>The documentation on this page is also still under construction. Not all
  features currently implemented have been completely documented.</p>
</div>
</div>

<p>For more information on the algorithms at the core of
  <strong>Monocle</strong>, or to learn more about how to use single-cell
  RNA-Seq to study complex biological processes, explore
  <a href= "{{ site.baseurl }}/papers/">our publications</a>.</p>

<p>Before we look at Monocle 3's functions for each of these common analysis
  tasks, let's see how to install Monocle.
</p>

<h2>Installing Monocle 3</h2>

<h3> Required software </h3>

<p>Monocle 3 runs in the <a href= "http://www.r-project.org/">R statistical
computing environment</a>. You will need R version 3.5 or higher,
<a href="http://bioconductor.org/">Bioconductor</a> version 3.5, and
monocle3 0.1.0 or higher to have access to the latest features.</p>

<p>To install Bioconductor, open R and run:</p>
{% highlight R %}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install()
{% endhighlight %}

<p>Next, install a few Bioconductor dependencies that aren't automatically installed:</p>
{% highlight R %}
BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats',
                       'limma', 'S4Vectors', 'SingleCellExperiment',
                       'SummarizedExperiment'))
{% endhighlight %}

<p>Optional: If you would like to use the resolution parameter in cluster_cells, you
  will also need to install one Python package. This can be done with the following code in R:</p>

  {% highlight R %}
install.packages("reticulate")
reticulate::py_install("louvain")
{% endhighlight %}


<p>Now, install monocle3 through the cole-trapnell-lab GitHub, execute:</p>

{% highlight R %}
devtools::install_github('cole-trapnell-lab/monocle3')
{% endhighlight %}

<h3> Testing the installation </h3>

<p> To ensure that Monocle 3 was installed correctly, start a new R session and
  run: </p>

{% highlight R %}
library(monocle3)
{% endhighlight %}

<h2>Getting Help</h2>

<p>Questions about Monocle 3 should be posted on our <a
href= "https://groups.google.com/d/forum/monocle-3-users">Google Group</a>.
Please do not email technical questions to Monocle contributors
directly.</p>

<!-- <h2>Recommended analysis protocol</h2>
<p>Monocle is a powerful toolkit for analyzing single-cell RNA-seq.  You don't need
to use all of its features for every analysis, and there are more than one way
to do some steps. The workflow is broken up into broad steps. When there's more than
one way to do a certain step, we've labeled the options as follows:
<table class= "table">
<tbody>
<tr>
<td><span class= "label label-danger">Required</span></td>
<td>You need to do this.</td>
</tr>
<tr>
<td><span class= "label label-info">Recommended</span> </td>
<td>Of the ways you could do this, we recommend you try this one first.</td>
</tr>
<tr>
<td><span class= "label label-success">Alternative</span></td>
<td>Of the ways you could do this, this way might work better than the one we usually recommend.</td>
</tr>
</tbody>
</table> -->

<h2>Workflow steps at a glance</h2>

<div class= "text-center">
  <img src= "{{site.baseurl}}/images/monocle3_new_workflow.png" width= 760>
</div>

<p>Below, you can see snippets of code that highlight the main steps of
  Monocle 3. Click on the section headers to jump to the detailed sections
  describing each one and allowing you to try the steps on example data. </p>

<h4><a href="{{ site.baseurl }}/monocle3_docs/#the-cell-data-set-class">Store data in a cell_data_set object</a></h4>
The first step in working with Monocle 3 is to load up your data into
Monocle 3's main class, <code><a href = "{{ site.baseurl }}/monocle3_docs/#the-cell-data-set-class">cell_data_set</a></code>:

{% highlight R %}
cds <- new_cell_data_set(expression_matrix,
                         cell_metadata = cell_metadata,
                         gene_metadata = gene_annotation)
{% endhighlight %}

<h4><a href= "{{ site.baseurl }}/monocle3_docs/#clustering-and-classifying-your-cells">Cluster your cells</a></h4>

You can easily cluster your cells to find new types:

{% highlight R %}
## Step 1: Normalize and pre-process the data
cds <- preprocess_cds(cds, num_dim = 100, residual_model_formula_str = "~ batch")

## Step 2: Reduce the dimensions using UMAP
cds <- reduce_dimension(cds)

## Step 3: Cluster the cells
cds <- cluster_cells(cds)
{% endhighlight %}

<h4><span class= "label label-info">Optional</span> <a href= "{{ site.baseurl }}/monocle3_docs/#constructing-single-cell-trajectories">Order cells in pseudotime along a trajectory</a> </h4>
Now, put your cells in order by how much progress they've made through whatever
process you're studying, such as differentiation, reprogramming, or an immune
response.

{% highlight R %}
## Step 4: Learn a graph
cds <- learn_graph(cds)

## Step 5: Order cells
cds <- order_cells(cds)

plot_cells(cds)
{% endhighlight %}

<h4><span class= "label label-info">Optional</span> <a href= "{{ site.baseurl }}/monocle3_docs/#differential-expression-analysis">Perform differential expression analysis</a> </h4>
Compare groups of cells in myriad ways to find differentially expressed genes,
controlling for batch effects and treatments as you like:

{% highlight R %}
# With regression:
gene_fits <- fit_models(cds, model_formula_str = "~embryo.time")
fit_coefs <- coefficient_table(gene_fits)
emb_time_terms <- fit_coefs %>% filter(term == "embryo.time")
emb_time_terms <- emb_time_terms %>% mutate(q_value = p.adjust(p_value))
sig_genes <- emb_time_terms %>% filter (q_value < 0.05) %>% pull(gene_short_name)

# With graph autocorrelation:
pr_test_res <- graph_test(cds,  neighbor_graph="principal_graph", cores=4)
pr_deg_ids <- row.names(subset(pr_test_res, q_value < 0.05))
{% endhighlight %}

<h2>Getting started with Monocle 3</h2>

Monocle 3 takes as input cell by gene expression matrix. Monocle 3 is designed
for use with absolute transcript counts (e.g. from UMI experiments). Monocle 3
works "out-of-the-box" with the transcript count matrices produced by
<a href= "https://support.10xgenomics.com/single-cell/software/overview/welcome">Cell Ranger</a>,
the software pipeline for analyzing experiments from the 10X Genomics Chromium
instrument. Monocle 3 also works well with data from other RNA-Seq workflows
such as <a href= "https://science.sciencemag.org/content/357/6352/661">sci-RNA-Seq</a>
and instruments like the Biorad ddSEQ.

<h3>The cell_data_set class</h3>

Monocle holds single-cell expression data in objects of the
<code>cell_data_set</code> class. The class is derived from the Bioconductor
<code><a href="http://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html">SingleCellExperiment</a></code> class,
which provides a common interface familiar to those who have analyzed other
single-cell experiments with Bioconductor. The class requires three input files:
<p>
<ul>
<li> <code>expression_matrix</code>, a numeric matrix of expression values,
  where rows are genes, and columns are cells </li>
<li> <code>cell_metadata</code>, a data frame, where rows are cells, and
  columns are cell attributes (such as cell type, culture condition, day
  captured, etc.)</li>
<li> <code>gene_metadata</code>, an data frame, where rows are features (e.g.
  genes), and columns are gene attributes, such as biotype, gc content,
  etc.</li>
</ul>
</p>
<p>
<div class= "panel panel-warning">
<div class= "panel-heading">

<h3 class= "panel-title" data-toc-skip>Required dimensions for input files</h3>

</div>
<div class= "panel-body">
The expression value matrix <strong>must</strong>:
<ul>
<li>have the same number of columns as the <code>cell_metadata</code> has
  rows.</li>
<li>have the same number of rows as the <code>gene_metadata</code> has
  rows.</li>
</ul>
Additionally:
<ul>
<li>row names of the <code>cell_metadata</code> object should match the column
  names of the expression matrix. </li>
<li>row names of the <code>gene_metadata</code> object should match row names
  of the expression matrix.</li>
<li> one of the columns of the <code>gene_metadata</code> should be named
  "gene_short_name", which represents the gene symbol or simple name (generally
  used for plotting) for each gene.</li>
</ul>

</div>
</div>
<h3>Generate a cell_data_set</h3>

<p>You can create a new <code>cell_data_set</code> (CDS) object as follows:</p>

<p>
{% highlight R %}
# Load the data
expression_matrix <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_expression.rds"))
cell_metadata <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_colData.rds"))
gene_annotation <- readRDS(url("http://staff.washington.edu/hpliner/data/cao_l2_rowData.rds"))

# Make the CDS object
cds <- new_cell_data_set(expression_matrix,
                                   cell_metadata = cell_metadata,
                                   gene_metadata = gene_annotation)
{% endhighlight %}
</p>

<h3>Generate a cell_data_set from 10X output</h3>
<p>To input data from 10X Genomics Cell Ranger, you can use the
  <code>load_cellranger_data</code> function:</p>

    <p><strong>Note:</strong> <code>load_cellranger_data</code> takes an argument <code>umi_cutoff</code>
    that determines how many reads a cell must have to be included. <strong>By default, this is set to 100</strong>.
    If you would like to include all cells, set <code>umi_cutoff</code to 0.<p>
{% highlight R %}
# Provide the path to the Cell Ranger output.
cds <- load_cellranger_data("~/Downloads/10x_data")
{% endhighlight %}


<h3 data-toc-text= "Working with large data sets">Working with large data sets</h3>

<p> Some single-cell RNA-Seq experiments report measurements from tens of
  thousands of cells or more. As instrumentation improves and costs drop,
  experiments will become ever larger and more complex, with many conditions,
  controls, and replicates. A matrix of expression data with 50,000 cells and a
  measurement for each of the 25,000+ genes in the human genome can take up a
  lot of memory. However, because current protocols typically don't capture all
  or even most of the mRNA molecules in each cell, many of the entries of
  expression matrices are zero. Using <i>sparse matrices</i> can help you work
  with huge datasets on a typical computer. We generally recommend the use of
  sparse matrices for most users, as it speeds up many computations even for
  more modestly sized datasets. </p>

<p> To work with your data in a sparse format, simply provide it to Monocle 3
  as a sparse matrix from the <code>Matrix</code> package: </p>

<p>
{% highlight R %}
cds <- new_cell_data_set(as(umi_matrix, "sparseMatrix"),
      cell_metadata = cell_metadata,
      gene_metadata = gene_metadata)
{% endhighlight %}
</p>

<p>
<div class= "panel panel-warning">
<div class= "panel-heading">
<h3 class= "panel-title" data-toc-skip>Don't accidentally convert to a dense expression matrix</h3>
</div>
<div class= "panel-body">
The output from a number of RNA-Seq pipelines, including Cell Ranger, is
already in a sparseMatrix format (e.g. MTX). If so, you should just pass it
directly to <code>new_cell_data_set</code> without first converting it to a
dense matrix (via <code>as.matrix()</code>, because that may exceed your
available memeory.
</div>
</div>
</p>

Monocle's sparse matrix support is provided by the <code>Matrix</code> package.
Other sparse matrix packages, such as <code>slam</code> or
<code>SparseM</code> are not supported.

<div class="btn-toolbar">
  <button class="button first-button" onClick="window.location.href = '{{site.baseurl}}/cluster/';">Next &raquo;</button>
</div>


</div>
</div>
</div>
