<!DOCTYPE html>
<html lang="{{ page.lang | default: site.lang | default: "en" }}">

{% include head.html %}

<body>
  {% include header.html %}

  <!-- Main jumbotron for a primary marketing message or call to action -->
  <div class="jumbotron" style="background-image: url(images/jumbo_pancreas.png); background-size: 100%;";>
    <div class="container">
      <div class="text-center">
        <h1>Monocle</h1>
        <p>An analysis toolkit for single-cell RNA-seq.</p>
        <p><a class="btn btn-warning btn-lg" href="{{ site.baseurl }}/docs/#installing-monocle" role="button">Get started&raquo;</a></p>
      </div>
    </div>
  </div>

  <div class="container">
    <!-- Example row of columns -->
    <div class="row">

      <div class="col-md-4">
        <div class="text-center">
          <img src="images/HSMM_trajectory.png" height=300 width=300></img>
          <h2>Pseudotime</h2>
          <p>Build single-cell trajectories with the software that introduced pseudotime. Find cell fate decisions and the genes regulated as they're made.</p>
          <p><a class="btn btn-default" href="{{ site.baseurl }}/docs/#constructing-single-cell-trajectories" role="button">View details &raquo;</a></p>
        </div>
      </div>

      <div class="col-md-4">
        <div class="text-center">
          <img src="images/worm_t-sne.png" height=300 width=300></img>
          <h2>Clustering</h2>
          <p>Group and classify your cells based on gene expression. Identify new cell types and states and the genes that distinguish them.</p>
          <p><a class="btn btn-default" href="{{ site.baseurl }}/docs/#classifying-and-counting-cells" role="button">View details &raquo;</a></p>
        </div>
      </div>
      <div class="col-md-4">
        <div class="text-center">
          <img src="images/HSMM_violins.png" height=300 width=300></img>
          <h2>Differential expression</h2>
          <p>Find genes that vary between cell types and states, over trajectories, or in response to perturbations using statistically robust, flexible differential analysis.</p>
          <p><a class="btn btn-default" href="{{ site.baseurl }}/docs/#differential-expression-analysis" role="button">View details &raquo;</a></p>
        </div>
      </div>
    </div>

    <!-- START THE FEATURETTES -->

    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7 col-md-push-5">
        <h2 class="featurette-heading"><a href = "/docs/#classifying-and-counting-cells"> Classify and count cells.</a></h2>
        <p class="lead">Single-cell RNA-Seq experiments allow you to discover (and possibly rare) subtypes of cells. Monocle helps you identify them.</p>

        <!-- <p>Most single-cell RNA-Seq protocols capture only a small percentage of the mRNA in each cell, which leads to so-called "drop outs", where mRNAs that are probably present just aren't detected. This makes it hard to determine how many cells of each type there are in your experiment. Monocle has a machine-learning based algorithm for classifying cells by type even in the face of missing data, and even when your cells have been treated in different ways or collected from varying conditions. You can use it to track population dynamics, measure responses to perturbation, and even discover new cell types.</p> -->

      </div>
      <div class="col-md-5 col-md-pull-7">
        <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/front_page_cluster.png" width=300 height=300 alt="Generic placeholder image">
      </div>
    </div>


    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7">
        <h2 class="featurette-heading"><a href = "{{site.baseurl}}/docs/#differential-expression-analysis">Compare cell populations through differential expression.</a></h2>
        <p class="lead">Single-cell RNA-Seq promises to unveil new cell types and new functional states for known ones. Monocle includes a sophisticated but easy-to-use system for differential expression.</p>
        <!-- <p>Monocle allows you to compare groups of cells to identify genes that differ between them, even after controlling for other variables in the experiment. For example, the plot above shows genes that differ between two conditions, but only one of them, LAG3, is **also** different between cell types, even after subtracting the other effects. You can use Monocle's differential expression system to zero in on the genes that are most relevant in your experiment.</p> -->

      </div>
      <div class="col-md-5">
        <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/example_T_cell_differential_genes.png" width=300 height=300 alt="Generic placeholder image">
      </div>
    </div>


    <hr class="featurette-divider">

    <div class="row featurette">
      <div class="col-md-7 col-md-push-5">
        <h2 class="featurette-heading"><a href = "{{site.baseurl}}/docs/#clustering-cells">Identify new marker genes.</a></h2>
        <p class="lead">Many researchers are using single-cell RNA-Seq to discover new cell types. Monocle can help you purify them or characterize them further by identifying key marker genes that you can use in follow up experiments such as immunofluorescence or flow sorting.

          <!-- <p>Monocle leverages its differential analysis procedures so you can be assured of the statistical robustness of your marker panel, leading to reliable downstream analysis.</p>
          <p> In the plot above, Monocle has identified the top 20 most cell-type specific genes for some of the major lymphocytes in peripheral blood.</p> -->

        </div>
        <div class="col-md-5 col-md-pull-7">
          <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/all_cells_marker_heatmap.png" height=300 width=300 alt="Generic placeholder image">
        </div>
      </div>


      <hr class="featurette-divider">

      <div class="row featurette">
        <div class="col-md-7">
          <h2 class="featurette-heading"><a href = "{{site.baseurl}}/docs/#constructing-single-cell-trajectories">Robustly track changes over (pseudo) time.</a></h2>
          <p class="lead">
            In development, disease, and throughout life, cells transition from one state to another.
            Monocle introduced the concept of pseudotime, which is a measure of how far a cell has moved through biological progress. </p>

            <!-- p>The plot above shows the transcriptional trajectory from developing <a href="http://science.sciencemag.org/content/350/6265/1251">olfactory neurons</a>. (Hanchate et al, Science, 2015).
              <p>Essentially, Monocle treats each cell as a separate time point in an ultra-high resolution time series. Instead of three or four time points, you get three or four thousand - one for each cell in the experiment. This resolution allows you distinguish genes regulated early in a complex cascade of transcriptional changes from those more downstream. For example, in the plot above you can see that Neurod1 is downregulated before cells enter an immature neuronal state, after which Gng8 is upregulated. As the cells reach maturity, Cnga2 becomes highly expressed.

                <p>Monocle 2 includes a vastly improved algorithm for learning trajectories and ordering cells. It's based on a recently introduced technique called [reverse graph embedding](http://dl.acm.org/citation.cfm?id=2783309) (Mao et al, KDD 15), a highly scalable nonlinear manifold learning technique. -->

                </div>
                <!-- <div class="col-md-5">
                  <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/neuron_mst.png" width=300 height=300 alt="Generic placeholder image">
                </div> -->

                <div class="col-md-5">
                  <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/neuron_pseudotime_changes.png" width=300 height=300 alt="Generic placeholder image">
                </div>

              </div>



    <hr class="featurette-divider">

              <div class="row featurette">
                <div class="col-md-7 col-md-push-5">
                  <h2 class="featurette-heading"><a href = "{{site.baseurl}}/docs/#analyzing-branches-in-single-cell-trajectories">Dissect cellular decisions with branch analysis.</a></h2>
                  <p class="lead">Single-cell trajectory analysis how cells
                    choose between one of several possible
                    end states. The new reconstruction
                     algorithms introduced in Monocle 2 can robustly reveal
                     branching trajectories, along with the genes
                     that cells use to navigate these decisions.

                    </div>
                    <div class="col-md-5 col-md-pull-7">
                      <img class="featurette-image img-responsive center-block" src="{{site.baseurl}}/images/branch_trajectory.png" height=300 width=300 alt="Generic placeholder image">
                    </div>

                  </div>

                  <hr class="featurette-divider">

                  <!-- /END THE FEATURETTES -->

                  <hr>
                  <!--
                  <ul class="post-list">
                  {% for post in site.posts %}
                  <li><article><a href="{{ site.baseurl }}{{ post.url }}">{{ post.title }} <span class="entry-date"><time datetime="{{ post.date | date_to_xmlschema }}">{{ post.date | date: "%B %d, %Y" }}</time></span></a></article></li>
                  {% endfor %}
                </ul> -->

                <footer>
                  <p>&copy; 2017 Cole Trapnell</p>
                  <small>Jumbotron image &copy; 2017 Douglas Melton. Used with permission. </small>

                </footer>
              </div> <!-- /container -->

              {% include footer.html %}

            </body>

            </html>
